### DISCUSSION TITLE
`useReadContract` infinite type instantiation error

### DISCUSSION THREAD

**Discussion by drake-nathan** [wevm/wagmi#3584] [General]
I'm converting from v1 to v2, and swapped a Wagmi cli `useReadContract` instance. After the fact, TS was throwing this error: "Type instantiation is excessively deep and possibly infinite." I've never seen this error before, any ideas?

<img width="525" alt="Screenshot 2024-02-08 at 9 05 52 PM" src="https://github.com/wevm/wagmi/assets/22569638/ab4631de-c723-4ba8-8989-7293625d9543">

**Comment by tmm** [comment]
Can provide a minimal reproduction? The screenshot isn't helpful.

**Comment by armanal** [comment]
### Error: Type instantiation is excessively deep and possibly infinite

using this example code provided in migration guide should reproduce the issue
[Migration Guide](https://wagmi.sh/react/guides/migrate-from-v1-to-v2#deprecated-usebalance-token-parameter)

```
import { useReadContracts } from 'wagmi' 
import { erc20Abi } from 'viem' 

const result = useReadContracts({ 
  allowFailure: false, 
  contracts: [ 
    { 
      address: '0x6B175474E89094C44Da98b954EedeAC495271d0F', 
      abi: erc20Abi, 
      functionName: 'balanceOf', 
      args: ['0x4557B18E779944BFE9d78A672452331C186a9f48'], 
    }, 
    { 
      address: '0x6B175474E89094C44Da98b954EedeAC495271d0F', 
      abi: erc20Abi, 
      functionName: 'decimals', 
    }, 
    { 
      address: '0x6B175474E89094C44Da98b954EedeAC495271d0F', 
      abi: erc20Abi, 
      functionName: 'symbol', 
    }, 
  ] 
}) 
```

here Typescript gives this error on contracts property: 
`Type instantiation is excessively deep and possibly infinite`

### Environment
```
"@types/node": "20.2.6",
"@types/react": "^18.2.10",
"@types/react-dom": "^18.2.19",
"typescript": "^5.1.3"
```

**Reply by tmm** [reply]
Works fine in this [TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMBK6CGATAYQgDsYocBjGVOAXzgDMoIQ4ByAdxwHMRg2AUKEixEcdFAoAmAAwBBAEbA6jZqzYA3YOhCCBFEqnhR0qZABt4AXhQZs+IqXJVUACgQC4cHOfMQOAGI4wObIJgBcjD4YADSecAZOlNSRANrxXh5e2d54eCaoqJFsMgAeAGwAQgCMAOwArAAstY0AogAcAJwynY0EjY0AIjid7QqdTa3oeOhy-RNStdV4MgFscTleOEqREtLyShubDMjEVMAkAHI4IOjFCj44Z+gA8gzrGdk4UNxFcKklUqNer1Wo1dqtWq1Tq9RqVAKtTp4WrtOTlWpSYFSADM2OqBGq7XKIwYjXabAAukcvLRqYhPls8gU-oCqnUmi0Ot1ev0hiMxhM2tNZvN6otlqsPptvDtxJJZIpgHSvCczjALsRrrdijMKKBolLsrTPllNvh8qYWWU2Q1mm0uj0+gNhqNxpNhXNGgslis1sqZcBdvKDkqGYxTucrjc7uxUABPEAKCDmQ00o5UgS0ACUAgEBQsMAAdHgcDAcAIAPQVzYAPQA-AIgA)

**Comment by on47sky** [comment]
Ensure you have `"strictNullChecks: true"` in your tsconfig.json. Without this, it doesn't work properly.

**Reply by tmm** [reply]
FYI `strictNullChecks` is `true` by default when `strict` is `true`.

**Reply by on47sky** [reply]
tmm, of course, you're right. However, I've seen projects with `strict: true` and modifiers that disable some strict rules (especially in migrating legacy projects).

My language isn't perfect. Under _"Without this, it doesn't work properly."_ I meant _"If `strictNullChecks` is `false` in config"_

 I mentioned `strictNullChecks` because it's the only rule from the strict group that lead to this TS error.
 
The more universal and correct answer is:
Ensure you have `strict: true` and no standalone rules from the "strict" group in tsconfig.

**Comment by valeriivanchev** [comment]
@tmm @on47sky I am also getting the same error, my tsconfig is the default vite one. I am generating hooks using the wagmi cli and get the same error. Could you help?

**Reply by tmm** [reply]
If you think there is bug, feel free to create an issue with a minimal reproduction.
